Fix and simplify WeatherOverlay

pchote opened the pull request:
Fixes #17757 and makes the code much more maintainable.

The only noticable change compared with the last release is that the game starts with the weather effect in force, instead of it fading in over the first couple of seconds. The fade in / (unused) fade out logic added a lot of unnecessary complexity, so I removed it to simplify the cleanup.

sorcerer86pt commented:
the weather effect fix is working:

[nuclear winter snow effect](https://drive.google.com/file/d/1rzbjPy_fnd20uewZda5O6n2W0k18_N2O/view?usp=sharing)

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -70,245 +71,177 @@ public class WeatherOverlayInfo : ITraitInfo, ILobbyCustomRulesIgnore
-					targetWindXOffset = info.WindLevels[currentWindIndex];
-				}
+				windUpdateCountdown = self.World.LocalRandom.Next(info.WindTick[0], info.WindTick[1]);
+				if (self.World.LocalRandom.Next(2) == 1 && targetWindStrengthIndex > 0)
```
While we're here it might be a good idea to change this to `if (targetWindStrengthIndex > 0 && self.World.LocalRandom.Next(2) == 1)` so we don't calculate unnecessary random values.

abcdefg30 reviewed (commented):
Two things I noticed:
- In the TD campaign missions and on Fort Lonestar the rain always seems to start out raining directly vertical. Imho it would be more polished if the initial direction wasn't fixed. That is not a regression compared to last release however, so you can ignore this point if it is too complicated/time consuming to do. (Just figured we might as well since you're basically rewriting the entire thing already.)
- On the RA map Snow Town you can observe a lot of erratic looking movement of the particles. It looks like they are vibrating. That was only barely noticeable on single particles on the release version.

penev92 commented:
I haven't finished my pass on the code, but I noticed something that I could reproduce on 3 different maps (including Fort Lonestar) - if the camera is below a certain point on the map no effects get rendered. Seems to be a fixed distance from the top, presumably tied to the viewport size.

pchote commented:
Adding fixup tag for @abcdefg30's polish points.

pchote committed:
Fix and simplify WeatherOverlay:

- Fix rendering issues
- Track particles in world pixels instead of screen pixels
- Removed un/underused fade in/out support
- Update wind once per tick instead of once per particle
- Make Particle struct readonly

pchote commented:
Updated to randomize the initial wind strength.

I believe the "vibration" is caused by quantizing the particle to the world pixel scale, which can't be fixed without breaking out additional render passes to allow us to draw full-resolution sprites below the shroud.

abcdefg30 committed:
Start with randomized wind strength

abcdefg30 reviewed (approved):
Rebased and applied fixups. Lgtm now.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/0fdb2579bb6cd5e04d9f3f49c7a68c59a87e65fd...89e34d62ded1811dfe3cdfa0602c78548f245f2f)

https://github.com/OpenRA/OpenRA/commit/89b5e34320d2051ccc472b06db25c085efd536ea
https://github.com/OpenRA/OpenRA/commit/ae0745b5427593ee8d8e2215e5bd43dff2516947
